storage: check lease/destroyStatus in replicate queue

tbg opened the pull request:
While working with #41028, we encountered a RESTORE that was stuck on an
AdminScatter. Looking more closely, we realized that the retry loop in
AdminScatter got stuck on a replica that had since been removed (but
was still referenced by the scatter). We'd repeatedly call
replicateQueue.processOneChange(repl) which would try to transition out
of a joint configuration, but would hit a descriptor changed error,
which was then retried, ad infinitum.

Put both a lease and destroy status check into processOneChange to
avoid this class of problem.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41433)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41433#-:-Lqgd2pLC7tb-0k403fe:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @tbg)

___
*[pkg/storage/replicate_queue.go, line 299 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41433#-Lqgc0WJAlu3JNtfnoVE:-Lqgc0WK3N53DJSkX315:b-q8rbsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4849ab289f601e84a604b1d720d9732e282bc344/pkg/storage/replicate_queue.go#L299)):*
> ```Go
> 	// perform this check, which could lead to infinite loops.
> 	if  _, err := repl.IsDestroyed(); err != nil  {
> 		return false,err
> ```

gofmt?
___
*[pkg/storage/replicate_queue.go, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41433#-LqgcPdV6ChryuRyOonF:-LqgcPdV6ChryuRyOonG:b7cx6iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4849ab289f601e84a604b1d720d9732e282bc344/pkg/storage/replicate_queue.go#L302)):*
> ```Go
> 	}
> 	desc, zone := repl.DescAndZone()
> 	if _, ok := repl.leaseGoodToGo(ctx); !ok {
> ```

It sort of seems like this call should check the destroy status but I understand that it doesn't but let's save such an optimization for a later time when we make all of the checks of destroy status more principled. See also https://github.com/cockroachdb/cockroach/issues/32268.


<!-- Sent from Reviewable.io -->

tbg commented:
TFTR!

bors r=ajwerner

tbg committed:
storage: check lease/destroyStatus in replicate queue

While working with #41028, we encountered a RESTORE that was stuck on an
AdminScatter. Looking more closely, we realized that the retry loop in
AdminScatter got stuck on a replica that had since been removed (but
was still referenced by the scatter). We'd repeatedly call
replicateQueue.processOneChange(repl) which would try to transition out
of a joint configuration, but would hit a descriptor changed error,
which was then retried, ad infinitum.

Put both a lease and destroy status check into processOneChange to
avoid this class of problem.

Release note: None

tbg commented:
Switched to redirectOnOrAcquireLease, leaseGoodToGo has false negatives (for example, always returns false for exp-based leases).

bors r=ajwerner
